zvariant/src/file_path.rs

```diff
@@ -0,0 +1,348 @@
+///    let path_buf: PathBuf = file_path.into();
+///    let string_lossy = Cow::Borrowed("/hello/world");
+///    let path_lossy = path_buf.to_string_lossy();
+///    assert_eq!(string_lossy, path_lossy);
+///
+///     // /hello/world/\\xC9blahblah
+///    let x = [
+///        47, 104, 101, 108, 108, 111, 47, 119, 111, 114, 108, 100, 47,
+///        92, 120, 67, 57, 98, 108, 97, 104, 98, 108, 97, 104, 0x0,
+///    ];
+///    let file_path = FilePath::new(Cow::Borrowed(CStr::from_bytes_with_nul(&x).unwrap()));
+///    let p2 = FilePath::from(PathBuf::from("/hello/world/\\xC9blahblah"));
+///    assert_eq!(p2, file_path);
+/// ```
+#[derive(Type, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[zvariant(signature = "ay")]
+pub struct FilePath<'f>(Cow<'f, CStr>);
+
+impl<'f> FilePath<'f> {
+    pub fn new(cow: Cow<'f, CStr>) -> Self {
+        Self(cow)
+    }
+}
+
+/// Chops a vector of bytes until a nul character,
+/// and if it (the nul) does not exist, it appends it at
+/// the end.
+///
+/// # Return:
+/// - _*Guaranteed*_ nul terminated vector of bytes
+fn chop_vec_with_nul(vec: &Vec<u8>) -> Vec<u8> {
+    let mut vec_with_nul: Vec<u8> = vec![];
+    for v in vec {
+        if v == &0x0 {
+            vec_with_nul.push(v.to_owned());
+            return vec_with_nul;
+        }
+        vec_with_nul.push(v.to_owned());
+    }
+    vec_with_nul.push(0x0);
+    return vec_with_nul;
+}
+
+/// Chops a slice of bytes until a nul character,
+/// and if it (the nul) does not exist, it appends it at
+/// the end.
+/// This method won't allocate/copy memory unless the nul
+/// byte does not exist in the `bytes` parameter
+///
+/// # Return:
+/// - _*Guaranteed*_ nul terminated slice of bytes
+fn chop_bytes_with_nul<'a>(bytes: &'a [u8]) -> &'a [u8] {
+    // if the slice already contains a nul byte
+    // we don't need to allocate memory
+    for (index, byte) in bytes.iter().enumerate() {
+        if byte == &0x0 {
+            return &bytes[0..index];
+        }
+    }
+    // otherwise we have no option but to copy to a new slice
+    // since we borrow bytes immutably
+    // SAFETY: see [`GlobalAlloc::alloc`].
+    let slice = unsafe {
+        // consider switching to https://doc.rust-lang.org/std/alloc/trait.Allocator.html#tymethod.allocate
+        // when it becames a part of the stable release
+        let ptr = std::alloc::alloc(Layout::new::<u8>());
+        std::slice::from_raw_parts_mut(ptr, bytes.len() + 1)
+    };
+    slice[..bytes.len()].copy_from_slice(bytes);
+    slice[bytes.len()] = 0x0;
+    return slice;
+}
+
+impl From<CString> for FilePath<'static> {
+    fn from(value: CString) -> Self {
+        return FilePath(Cow::Owned(value));
+    }
+}
+
+impl<'f> From<&'f OsStr> for FilePath<'f> {
+    fn from(value: &'f OsStr) -> Self {
+        // SAFETY: this call is safe because we guarentee the nul termination
+        // of the `path_bytes`
+        let bytes_with_nul = chop_bytes_with_nul(value.as_encoded_bytes());
```

Comment:
* The safety comment is above the safe line.
* We could just use `CStr::from_bytes_with_null` and error out if null is missing. Then we don't need chop_bytes_with_null or unsafe even.